Sort L0 files by newly introduced epoch_num

hx235 opened the pull request:
**Context:**
Sorting L0 files by `largest_seqno` has at least two inconvenience:
-  File ingestion and compaction involving ingested files can create files of overlapping seqno range with the existing files. `force_consistency_check=true` will catch such overlap seqno range even those harmless overlap.
    - For example, consider the following sequence of events ("key@n" indicates key at seqno "n")
       - insert k1@1 to memtable m1
       - ingest file s1 with k2@2, ingest file s2 with k3@3
        - insert k4@4 to m1
       - compact files s1, s2 and  result in new file s3 of seqno range [2, 3]
       - flush m1 and result in new file s4 of seqno range [1, 4]. And `force_consistency_check=true` will think s4 and s3 has file reordering corruption that might cause retuning an old value of k1
    - However such caught corruption is a false positive since s1, s2 will not have overlapped keys with k1 or whatever inserted into m1 before ingest file s1 by the requirement of file ingestion (otherwise the m1 will be flushed first before any of the file ingestion completes). Therefore there in fact isn't any file reordering corruption.
- Single delete can decrease a file's largest seqno and ordering by `largest_seqno` can introduce a wrong ordering hence file reordering corruption
    - For example, consider the following sequence of events ("key@n" indicates key at seqno "n", Credit to @ajkr  for this example)
        - an existing SST s1 contains only k1@1
        - insert k1@2 to memtable m1
        - ingest file s2 with k3@3, ingest file s3 with k4@4
        - insert single delete k5@5 in m1
        - flush m1 and result in new file s4 of seqno range [2, 5]
        - compact s1, s2, s3 and result in new file s5 of seqno range [1, 4]
        - compact s4 and result in new file s6 of seqno range [2] due to single delete
    - By the last step, we have file ordering by largest seqno (">" means "newer") : s5 > s6 while s6 contains a newer version of the k1's value (i.e, k1@2) than s5, which is a real reordering corruption. While this can be caught by `force_consistency_check=true`, there isn't a good way to prevent this from happening if ordering by `largest_seqno`
   
Therefore, we are redesigning the sorting criteria of L0 files and avoid above inconvenience. Credit to @ajkr , we now introduce `epoch_num` which describes the order of a file being flushed or ingested/imported (compaction output file will has the minimum `epoch_num` among input files'). This will avoid the above inconvenience in the following ways: 
- In the first case above, there will no longer be overlap seqno range check in `force_consistency_check=true` but `epoch_number`  ordering check. This will result in file ordering s1 <  s2 <  s4 (pre-compaction) and s3 < s4 (post-compaction) which won't trigger false positive corruption. See test class `DBCompactionTestL0FilesMisorderCorruption*` for more.
- In the second case above, this will result in file ordering s1 < s2 < s3 < s4 (pre-compacting s1, s2, s3), s5 < s4 (post-compacting s1, s2, s3), s5 < s6 (post-compacting s4), which are correct file ordering without causing any corruption.

**Summary:**
- Introduce `epoch_number` stored per `ColumnFamilyData` and sort CF's L0 files by their assigned `epoch_number` instead of `largest_seqno`.
  - `epoch_number` is increased and assigned upon `VersionEdit::AddFile()` for flush (or similarly for WriteLevel0TableForRecovery) and file ingestion (except for allow_behind_true, which will always get assigned as the `kReservedEpochNumberForFileIngestedBehind`)
  - Compaction output file  is assigned with the minimum `epoch_number` among input files'
      - Refit level: reuse refitted file's epoch_number
  -  Other paths needing `epoch_number` treatment:
     - Import column families: reuse file's epoch_number if exists. If not, assign one based on `NewestFirstBySeqNo`
     - Repair: reuse file's epoch_number if exists. If not, assign one based on `NewestFirstBySeqNo`.
  -  Assigning new epoch_number to a file and adding this file to LSM tree should be atomic. This is guaranteed by us assigning epoch_number right upon `VersionEdit::AddFile()` where this version edit will be apply to LSM tree shape right after by holding the db mutex (e.g, flush, file ingestion, import column family) or  by there is only 1 ongoing edit per CF (e.g, WriteLevel0TableForRecovery, Repair). 
  - Assigning the minimum input epoch number to compaction output file won't misorder L0 files (even through later `Refit(target_level=0)`). It's due to for every key "k" in the input range, a legit compaction will cover a continuous epoch number range of that key. As long as we assign the key "k" the minimum input epoch number, it won't become newer or older than the versions of this key that aren't included in this compaction hence no misorder.
- Persist `epoch_number` of each file in manifest and recover `epoch_number` on db recovery
   - Backward compatibility with old db without `epoch_number` support is guaranteed by assigning `epoch_number` to recovered files by `NewestFirstBySeqno` order. See `VersionStorageInfo::RecoverEpochNumbers()` for more
   - Forward compatibility with manifest is guaranteed by flexibility of `NewFileCustomTag`
- Replace `force_consistent_check` on L0 with `epoch_number` and remove false positive check like case 1 with `largest_seqno` above
   - Due to backward compatibility issue, we might encounter files with missing epoch number at the beginning of db recovery. We will still use old L0 sorting mechanism (`NewestFirstBySeqno`) to check/sort them till we infer their epoch number. See usages of `EpochNumberRequirement`.
- Remove fix https://github.com/facebook/rocksdb/pull/5958#issue-511150930 and their outdated tests to file reordering corruption because such fix can be replaced by this PR.
- Misc: 
   - update existing tests with `epoch_number` so make check will pass
   - update https://github.com/facebook/rocksdb/pull/5958#issue-511150930 tests to verify corruption is fixed using `epoch_number` and cover universal/fifo compaction/CompactRange/CompactFile cases
   - assert db_mutex is held for a few places before calling ColumnFamilyData::NewEpochNumber()

**Test:**
- `make check`
- New unit tests under `db/db_compaction_test.cc`, `db/db_test2.cc`, `db/version_builder_test.cc`, `db/repair_test.cc`
- Updated tests (i.e, `DBCompactionTestL0FilesMisorderCorruption*`) under https://github.com/facebook/rocksdb/pull/5958#issue-511150930
- [Ongoing] Compatibility test: manually run https://github.com/ajkr/rocksdb/commit/36a5686ec012f35a4371e409aa85c404ca1c210d (with file ingestion off for running the `.orig` binary to prevent this bug affecting upgrade/downgrade formality checking) for 1 hour on `simple black/white box`, `cf_consistency/txn/enable_ts with whitebox + test_best_efforts_recovery with blackbox`
- [Ongoing] normal db stress test
- [Ongoing] db stress test with aggressive value https://github.com/facebook/rocksdb/pull/10761

hx235 commented on db/db_impl/db_impl_compaction_flush.cc:
```diff
@@ -1697,13 +1687,20 @@ Status DBImpl::ReFitLevel(ColumnFamilyData* cfd, int level, int target_level) {
           f->marked_for_compaction, f->temperature, f->oldest_blob_file_number,
-          f->oldest_ancester_time, f->file_creation_time, f->file_checksum,
+          f->oldest_ancester_time, f->file_creation_time,
+          to_level == 0 ? l0_epoch_num-- : f->l0_epoch_number, f->file_checksum,
```
I think it should be fine that we assign l0_epoch_num based on `vstorage->LevelFiles(level)`'s order (i.e, smallest key in non-zero level) since files should not have overlapped keys. Correct me if I am missing anything.

hx235 commented:
I need some help in debugging `
ci/circleci: build-linux-encrypted_env-no_compression —` which I can't repro on my devserver nor on the CI server but consistently fail. I will try printing statement to find out which is exactly calling GetFileSize() problematically but let me know if there is a smarter way of figuring this out.

hx235 commented:
Update: update the DBTest2.SortL0FilesByL0EpochNumber to use consistent option in the test to mitigate `ci/circleci: build-linux-encrypted_env-no_compression ` error https://app.circleci.com/pipelines/github/facebook/rocksdb/21041/workflows/7ff39cf3-49b3-468f-8df4-9208c8dd4f1d/jobs/528240

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6278,248 +6273,534 @@ void IngestOneKeyValue(DBImpl* db, const std::string& key,
+  std::atomic<bool> compaction_path_sync_point_called_;
+};
+
+TEST_F(DBCompactionTestL0FilesReorderingCorruption,
```
I decided to replace original `DBCompactionTestWithParam.IntraL0CompactionAfterFlushCheckConsistencyFail` and `DBCompactionTestFIFOCheckConsistencyWithParam.FlushAfterIntraL0CompactionWithIngestedFile` with one test case here cuz differentiating by memtable being sealed or not does not seem quite important for L0 epoch number approach.

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6278,248 +6273,534 @@ void IngestOneKeyValue(DBImpl* db, const std::string& key,
+  sleeping_tasks.WaitUntilDone();
+  ASSERT_OK(Flush());
+  ASSERT_EQ(2, NumTableFilesAtLevel(0));
+  SequenceNumber flushed_file_largest_seqno =
```
[a minor TODO] Now that I look at this again, comparing largest_seqno might not be needed or done in a better way with less assumption like "GetLatestL0FileLargestSeqnoHelper(). Pretty minor so I will address this along with your other feedback.

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6278,248 +6273,534 @@ void IngestOneKeyValue(DBImpl* db, const std::string& key,
   Compact("", Key(99));
   ASSERT_EQ(0, NumTableFilesAtLevel(0));
 
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
```
[a minor TODO] Now I realize I should have encapsulated the sync point into a function like I did for Universal/FIFO case (probably also for stop/resume flush job too). Pretty minor so will address this along with your other feedback.

hx235 commented:
Update @ajkr 
will be updating ImportColumnFamily to consider l0_epoch_number of the imported file and RepairDB to solve some edge cases 

will be updating the test for readability too

hx235 commented on db/version_builder.cc:
```diff
@@ -32,26 +32,22 @@
+      if (lhs->l0_epoch_number != rhs->l0_epoch_number) {
+        return lhs->l0_epoch_number > rhs->l0_epoch_number;
+      } else {
+        return false;
```
[Todo] Replace this as there can be legit echo_number tie due to multiple compaction output files

hx235 commented on db/external_sst_file_ingestion_job.cc:
```diff
@@ -469,8 +469,10 @@ Status ExternalSstFileIngestionJob::Run() {
-        oldest_ancester_time, current_time, f.file_checksum,
-        f.file_checksum_func_name, f.unique_id);
+        oldest_ancester_time, current_time,
+        f.picked_level == 0 ? versions_->NewL0EpochNumber()
```
[TODO] Should assign epoch_number based on seqno within a bulk of ingested files.

hx235 commented:
Update: extend l0_epoch_num to epoch_num 
Next step:
- extend UT
- stress testing
- PR summary

hx235 commented:
Update:
- extended test from l0_epoch_number to epoch_number
- extended UT coverage for epoch_number 
- misc

Next step:
- Repair DB + file ingestion 
- Rerun stress test + formality test
- Update PR summary

hx235 commented:
Update:
- well, it turned out that pre-existing issue of file ingestion will make it seen as corrupted when RepairDB and get skipped. Therefore postpone RepairDB + file ingestion 


Next step:
- Run formality test + stress testing 
- Update PR summary

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6278,327 +6273,709 @@ void IngestOneKeyValue(DBImpl* db, const std::string& key,
+  std::shared_ptr<test::SleepingBackgroundTask> sleeping_task_;
+};
+
+TEST_F(DBCompactionTestL0FilesMisorderCorruption,
```
[Note] I combine the previous level compaction's test FlushAfterIntraL0CompactionCheckConsistencyFail and IntraL0CompactionAfterFlushCheckConsistencyFail cuz one test is enough to make the point now.

hx235 commented on db/external_sst_file_ingestion_job.cc:
```diff
@@ -469,8 +469,8 @@ Status ExternalSstFileIngestionJob::Run() {
         f.assigned_seqno, false, f.file_temperature, kInvalidBlobFileNumber,
-        oldest_ancester_time, current_time, f.file_checksum,
-        f.file_checksum_func_name, f.unique_id);
+        oldest_ancester_time, current_time, versions_->NewEpochNumber(),
```
[Note] I tried out that IngestExternalFile() can ingest multiple files with the same keys without complain. So here I decided to assign different epoch number. Should be fine according the import order as the seqno/file number is assigned by the import order.

hx235 commented on db/repair.cc:
```diff
@@ -215,12 +215,21 @@ class Repairer {
       ExtractMetaData();
 
+      // Recover epoch_number of existing SST files and next epoch_number
+      if (HasMissingEpochNumber()) {
```
[Note]
Unfortunately, I couldn't get RepairDB() fails on ingested file cuz those files are seen as corrupted due to https://github.com/facebook/rocksdb/blob/7.7.fb/db/repair.cc#L546-L555 and hence are skipped.

ImportColumnFamily also didn't work with RepairDB cuz https://github.com/facebook/rocksdb/blob/7.7.fb/db/repair.cc#L524 will returned the exporting CF id instead of the created CF id. Then we will have same files of same epoch number under the same CF, violating my added check https://github.com/facebook/rocksdb/pull/10922/files#diff-a6aa385940793f95a2c5b39cc670bd440c4547fa54fd44622f756382d5e47e43R422

hx235 commented:
Update:
- running format testing
- figure out the undef ref to static const tmr and run stress run

hx235 commented:
@ajkr Ready for review - compatibility tests/normal stress run seems fine. Stress testing with aggressive CI value causes some verification errors that I can't conclude whether it's an issue with this PR or stress testing framework with epoch_number. More info in offline chat.

Nerotool commented:
SEND A DM TO NEROTOOL_ ON INSTAGRAM FOR HACKING

hx235 commented:
Head-up: due to https://github.com/facebook/rocksdb/pull/10999, this PR will need a rebase (if pr10999) is adopted.

hx235 commented:
Update:
- Rebase
- Modified FIFO related UT a bit since PickCompactionsToWarm() do nothing when num_level = 1 after https://github.com/facebook/rocksdb/commit/e267909ecfe92e79232f7700e41b6e12c4a87c01

ajkr reviewed (commented):
Some comments. I'm not done thinking about compatibility, import, and repair.

hx235 commented on db/flush_job.cc:
```diff
@@ -912,6 +913,7 @@ Status FlushJob::WriteLevel0Table() {
           &oldest_ancester_time);
       meta_.oldest_ancester_time = oldest_ancester_time;
       meta_.file_creation_time = current_time;
+      meta_.epoch_number = versions_->NewEpochNumber();
```
Ah no wonder the `meta_.fd = FileDescriptor(versions_->NewFileNumber(), 0, 0);` is put somewhere else instead of here. Good catch.

hx235 replied:
Fixed

hx235 replied:
Also update the PR summary on how do we guarantee assigning epoch number to a file and adding to file to LSM tree is atomic across all epoch number treatment.

hx235 replied:
My previous fix is not good enough - see my latest one.

ajkr replied:
What was the issue with doing this in the same lock hold as `PickMemTable()`? If we want to ever include this in table properties, we will need to choose the epoch number before `BuildTable()`.

hx235 replied:
Oh nvm - I realized in `TryInstallMemtableFlushResults()`, it guarantees the earlier memtable's edit get processed first thru
```
 if (!m->flush_completed_) {
        break;
      }
```

I was originally concerned with the second parallel flush job with larger epoch number finishes first and get its edit applied to LSM tree. Will revert this.

hx235 commented on db/compaction/compaction_picker.cc:
```diff
@@ -31,8 +31,7 @@ bool FindIntraL0Compaction(const std::vector<FileMetaData*>& level_files,
-                           SequenceNumber earliest_mem_seqno) {
+                           CompactionInputFiles* comp_inputs) {
   // Do not pick ingested file when there is at least one memtable not flushed
   // which of seqno is overlap with the sst.
```
Fixed

hx235 commented on HISTORY.md:
```diff
@@ -18,6 +18,7 @@
 * Fixed a regression in scan for async_io. During seek, valid buffers were getting cleared causing a regression.
 * Tiered Storage: fixed excessive keys written to penultimate level in non-debug builds.
 * Fixed a bug that multi-level FIFO compaction deletes one file in non-L0 even when `CompactionOptionsFIFO::max_table_files_size` is no exceeded since #10348 or 7.8.0.
+* Fix L0 file misorder corruption caused by ingesting files of overlapping seqnos with memtable entries' through introducing `epoch_number`. Before the fix, `force_consistency_checks=true` may catch the corruption before it's exposed to readers, in which case writes returning `Status::Corruption` would be expected. Also replace the previous incomplete fix (#5958) to the same corruption with this  ...
```
Fixed - made this mistake again!!!

hx235 commented on db/compaction/compaction_service_job.cc:
```diff
@@ -190,6 +190,7 @@ CompactionJob::ProcessKeyValueCompactionWithCompactionService(
     meta.largest.DecodeFrom(file.largest_internal_key);
     meta.oldest_ancester_time = file.oldest_ancester_time;
     meta.file_creation_time = file.file_creation_time;
+    meta.epoch_number = compaction->MinInputFileEpochNumber();
```
Fixed

hx235 commented on db/db_compaction_test.cc:
```diff
@@ -6271,172 +6273,655 @@ void IngestOneKeyValue(DBImpl* db, const std::string& key,
+
+    if (compaction_path_to_test == "PickPeriodicCompaction") {
+      AddFilesMarkedForPeriodicCompaction(5);
+    } else {
```
Fixed

hx235 commented on db/version_builder.cc:
```diff
@@ -382,42 +381,65 @@ class VersionBuilder::Rep {
+            } else if (icmp->Compare(lhs->smallest, rhs->smallest) > 0 &&
+                       icmp->Compare(rhs->largest, lhs->smallest) >= 0) {
+              range_overlapped = true;
+            }
```
Fixed

hx235 commented:
Update: addressed current comments + remove epoch_number from TableProperties and add WARNING on regenerating epoch_number as discussed. 

Going thru stress test again and will let you know here when finished.

hx235 commented:
Update: @ajkr 
I tried out the second approach (VersionBuilder check conditioned on VersionStorageInfo). Basically, I am passing down `EpochNumberRequirement` from VersionSet::Recover() to versions created within VersionEditHandler::Iterate(). So later version builder can look at VersionStorageInfo to determine how to sort and check its L0 files. Fixup is done on the latest VersionStorageInfo of each CF after VersionEditHandler::Iterate(). 

Awkward thing is about how to re-sort the ths VersionStorage's files (previously sorted by seqno) now by epoch number after we fix up the epoch number. I noticed that VersionStorage's files are sorted by VersionBuilder::SaveTo() and SaveTo() has consequence in VersionStorageInfo::PrepareForVersionAppend(). So I have to delete all the file, add them back in dummy builders, save to a new version and make this new version the latest. Not sure if there is better way to do it.

I tried out approach (3) a bit yesterday - wasn't satisfying. Biggest challenges lies in (1) how to craft out recovery-related epoch logic out of VersionEditHandler if we keep the actual version creation within VersionEditHandler (2) If we decide to move it out of VersionEditHandler, figure out the impact of callers of VersionEditHandler::CheckIterationResults. 

But if we are not satisfied by the overall approach of this PR (approach (2)), I wouldn't mind trying (3) again.  After all, 24 hours ago, I thought approach (2) was more messy than I see now.

hx235 commented on db/version_set.cc:
```diff
@@ -5721,7 +5755,177 @@ Status VersionSet::TryRecoverFromOneManifest(
+  s = del_builder->Apply(&file_deletions);
+  if (s.ok()) {
+    auto* v = new Version(
+        cfd, this, file_options_, *cfd->GetLatestMutableCFOptions(), io_tracer_,
```
[TODO] Use EpochNumberRequirement::kMightMissing here and eliminate InTransition enum

hx235 replied:
Fixed

hx235 commented on db/version_set.cc:
```diff
@@ -5721,7 +5755,177 @@ Status VersionSet::TryRecoverFromOneManifest(
+
+  if (s.ok()) {
+    auto* v = new Version(
+        cfd, this, file_options_, *cfd->GetLatestMutableCFOptions(), io_tracer_,
```
[TODO] Use kMustPresent here instead

hx235 replied:
Fixed

hx235 commented on db/version_set.cc:
```diff
@@ -5721,7 +5755,177 @@ Status VersionSet::TryRecoverFromOneManifest(
+      ROCKS_LOG_WARN(db_options()->info_log,
+                     "CF(%d)'s epoch numbers are inferred based on seqno",
+                     cf_id);
+      s = InstallInferredEpochNumberResult(cf_id);
```
Right - I was somehow mistakenly thinking about the case of needing to re-sort by inferred epoch number and have all these complexity and overlooked the fact that this was already sorted by seqno (that's inferred epoch number)

hx235 replied:
Fixed.

hx235 commented on db/version_builder.cc:
```diff
@@ -382,43 +365,79 @@ class VersionBuilder::Rep {
-            std::ostringstream oss;
-            oss << "L0 files are not sorted properly: files #"
-                << lhs->fd.GetNumber() << ", #" << rhs->fd.GetNumber();
+          if (epoch_number_requirement ==
+              EpochNumberRequirement::kMightMissing) {
+            if (!level_zero_cmp_by_seqno_(lhs, rhs)) {
+              std::ostringstream oss;
+              oss << "L0 files are not sorted properly: files #"
+                  << lhs->fd.GetNumber() << " with seqnos (largest, smallest) "
+                  << lhs->fd.largest_seqno << " , " << lhs->fd.smallest_seqno
+                  << ", #" << rhs->fd.GetNumber()
+                  << " with seqnos (largest, smallest) "
+                  << rhs->fd.largest_seqno << " , " << rhs->fd.smallest_seqno;
+              return Status::Corruption("VersionBuilder", oss.str());
```
Yes! Good idea

hx235 replied:
Fixed

hx235 commented on db/version_set.cc:
```diff
@@ -5721,7 +5755,177 @@ Status VersionSet::TryRecoverFromOneManifest(
+}
+
+Status VersionSet::RecoverEpochNumbers() {
+  std::unordered_map<uint32_t, bool> cf_to_has_missing_epoch_number;
+
+  PopulateCfToHasMissingEpochNumber(cf_to_has_missing_epoch_number);
+  UpdateNextEpochNumberBasedOnPresentEpochNumber(
+      cf_to_has_missing_epoch_number);
+  Status s = InferMissingEpochNumber(cf_to_has_missing_epoch_number);
```
Okay - let met try with moving it to CFD

hx235 commented on db/version_set.cc:
```diff
@@ -5721,7 +5755,177 @@ Status VersionSet::TryRecoverFromOneManifest(
+          files.push_back(f);
+        }
+      }
+      InferEpochNumbersFromSeqNo(files);
```
Fixed, also for import CF.

hx235 commented:
Update @ajkr:
- Addressed comments
- Simplified a bit on RepairDB-related epoch logic
- Ongoing compatibility check/stress test and will notify here when finish.

hx235 commented:
@ajkr  ready for review again.

hx235 commented:
@ajkr Ready for review (just addressed some minor lint warning)

ajkr commented on db/version_builder.cc:
```diff
@@ -382,43 +365,76 @@ class VersionBuilder::Rep {
+                     EpochNumberRequirement::kMustPresent) {
+            if (lhs->epoch_number == kUnknownEpochNumber) {
+              std::ostringstream oss;
+              oss << "L0 file is not assigned with valid epoch number: files #"
+                  << lhs->fd.GetNumber() << " with epoch number "
+                  << lhs->epoch_number;
+              return Status::Corruption("VersionBuilder", oss.str());
+            }
 
-            return Status::Corruption("VersionBuilder", oss.str());
-          }
+            if (rhs->epoch_number == kUnknownEpochNumber) {
+              std::ostringstream oss;
+              oss << "L0 file is not assigned with valid epoch number: files #"
+                  << rhs->fd.GetNumber() << " with epoch number "
+                  << rhs->epoch_number;
+              return Status::Corruption("VersionBuilder", oss.str());
+            }
```
As mentioned in another comment, there may be a better place (write side) for this validation.

hx235 replied:
Fixed by removing this kUnknownEpochNumber and revised the related UT.

ajkr reviewed (changes requested):
Finally finished thinking about all the parts I hoped to think about. Great work on this very difficult problem

hx235 commented on db/import_column_family_job.cc:
```diff
@@ -307,6 +310,70 @@ Status ImportColumnFamilyJob::GetIngestedFileInfo(
+
+  for (int level = static_cast<int>(files_by_level.size()) - 1; level >= 1;
+       --level) {
+    auto& files = files_by_level[level];
+    if (files.empty()) {
+      continue;
+    }
+    uint64_t next_epoch_number = cfd_->NewEpochNumber();
+    for (LiveFileMetaData* f : files) {
+      f->epoch_number = next_epoch_number;
+    }
+  }
+
+  NewestFirstBySeqNo cmp;
+  assert(files_by_level.size() >= 1);
+  std::sort(files_by_level[0].begin(), files_by_level[0].end(), cmp);
+  for (auto iter = files_by_level[0].rbegin(); iter != files_by_level[0].rend();
+       iter++) {
+    LiveFileMetaData* f = *iter;
+    f->epoch_number = cfd_->NewEpochNumber();
+  }
+
+  ROCKS_LOG_WARN(versions_->db_options()->info_log,
+                 "Imported CF(%d)'s epoch numbers are inferred based on seqno",
+                 cfd_->GetID());
+}
+
+void ImportColumnFamilyJob::RecoverEpochNumbers() {
+  if (HasMissingEpochNumber()) {
+    InferMissingEpochNumber();
+  } else {
+    uint64_t max_epoch_number = kUnknownEpochNumber;
+    for (std::size_t i = 0; i < metadata_.size(); ++i) {
+      const LiveFileMetaData& live_file_metadata = metadata_[i];
+      max_epoch_number =
+          std::max(max_epoch_number, live_file_metadata.epoch_number);
+    }
+    cfd_->SetNextEpochNumber(max_epoch_number + 1);
+  }
+}
```
Thanks! Approved and will rebase on yours.

hx235 replied:
Fixed and rebased.

hx235 commented on db/column_family.cc:
```diff
@@ -1520,6 +1518,42 @@ FSDirectory* ColumnFamilyData::GetDataDir(size_t path_id) const {
+    }
+
+    ROCKS_LOG_WARN(ioptions_.info_log.get(),
+                   "CF(%d)'s epoch numbers are inferred based on seqno", id_);
```
Fixed

hx235 commented on db/external_sst_file_ingestion_job.cc:
```diff
@@ -469,8 +469,8 @@ Status ExternalSstFileIngestionJob::Run() {
         f.assigned_seqno, false, f.file_temperature, kInvalidBlobFileNumber,
-        oldest_ancester_time, current_time, f.file_checksum,
-        f.file_checksum_func_name, f.unique_id);
+        oldest_ancester_time, current_time, cfd_->NewEpochNumber(),
```
Oh I am embarrassed to miss such important feature in file ingestion. Fixed it and add UT. Also open another PR https://github.com/facebook/rocksdb/pull/11033 (so unexpected blockers there does not block this PR) to add stress-test support that could've notified me earlier.

hx235 replied:
Update: has one more thing to fix for this!

hx235 commented on db/column_family.cc:
```diff
@@ -1520,6 +1518,42 @@ FSDirectory* ColumnFamilyData::GetDataDir(size_t path_id) const {
+
+    ROCKS_LOG_WARN(ioptions_.info_log.get(),
+                   "CF(%d)'s epoch numbers are inferred based on seqno", id_);
+    vstorage->SetEpochNumberRequirement(EpochNumberRequirement::kMustPresent);
```
> we should reconsider where to do the kUnknownEpochNumber validation

Fixed. Picked "the VersionEdit::EncodeTo() could return false for invalid" so the dummy vstorage approach in CF import/Repair could work without violating assertion in  VersionEdit::AddFile()

hx235 commented:
Update: Addressed the comments and going thru compatibility + stress test again. Will notify here when it's ready for review again.

ajkr reviewed (approved):
Thanks for addressing all the comments. LGTM!!

hx235 commented:
> Thanks for addressing all the comments. LGTM!!

Thanks for the review!!!  Compatibility test and stress tests finish good so I will land this once signals are clear.
